src/preamble/02_what_is_a_module_and_a_package.md

```diff
@@ -1 +1,246 @@
-# What is a module and a package?
+# What is a module, a component, and a package?
+
+## TL;DR
+
+A **module** is a unit of code defined by GHC and combined by cabal to create a
+**component**. A **package** is a cabal project, usually denoted by a
```

Comment:
Actually, a cabal 'project' is probably the thing in which you have a `cabal.project` file and one or more subdirectories with cabal packages (each having `foo.cabal` file). If only one package, no subdirectory is needed.

Reply:
Yeah, I figured there would be ambiguity here. Do you have a good word for what a package is that is not self descriptive? Perhaps a "code repository that represents a collection of components"?

Reply:
As an aside it is really hard to be precise, and provide a definition that doesn't use the word you are defining, while also avoiding name collisions with other concepts! lol.

Reply:
Yes, mission impossible. :D

You may want to try and systematize your approach. E.g., go from small to large, defining large from the small. Something like "module is ...; package is a collection of modules contained in a directory and organized into components using a package description file (e.g., `foo.cabal` [use whatever is your running example]); project is a collection of one or more packages configured with a `cabal.project` file residing in the topmost directory of the project", etc.,

Reply:
+1 for going from small to large. Perhaps taking some inspiration from the introduction of http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.113.4699&rep=rep1&type=pdf

Random thoughts:
1. modules: separate compilation (.o per module), namespace management, abstraction (module signatures)
2. library components: code reuse, separate compilation (.a/.so/.dll per library), namespace management (not exposed modules), abstraction (module instantiation)
3. packages: code reuse, namespace management (private library components), grouping with related programs and testsuites, unit of distribution (e.g. on Hackage), loose dependencies on other packages
4. projects: grouping several related packages, less loose/fixed dependencies on other packages (freeze files), not distributed except via repository. Two independent project manager tools: stack and cabal-install

Reply:
I think I want to defer discussion of `cabal.project` to the designated section, but I agree with all other points.